src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,330 @@
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+    // We received arguments from the client, deserialize them back to JS.
+    if (serializedArgs.size() > 0) {
+      auto args = KJ_REQUIRE_NONNULL(
+          deserializeV8(js, serializedArgs).tryCast<jsg::JsArray>(),
+          "expected JsArray when deserializing arguments.");
+      // Call() expects a `Local<Value> []`... so we populate an array.
+      auto arguments = kj::heapArray<v8::Local<v8::Value>>(args.size());
+      for (size_t i = 0; i < args.size(); ++i) {
+        arguments[i] = args.get(js, i);
+      }
+      return jsg::check(fn->Call(js.v8Context(), fn, args.size(), arguments.begin()));
+    } else {
+      return jsg::check(fn->Call(js.v8Context(), fn, 0, nullptr));
+    }
+  };
+
+  struct Returned {
+    kj::Array<kj::byte> serializedValue;
+  };
+  struct ThrewException {
+    kj::Array<kj::byte> serializedException;
+  };
+  // The serialized JS value.
+  using SerializedResult = kj::OneOf<Returned, ThrewException>;
+
+  // Invokes the given function and serializes the result for RPC.
+  jsg::Promise<SerializedResult> callFuncAndSerialize(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    // `processAsyncResult` expects a promise, so even if we return synchronously we want to wrap
+    // the result of calling the JS function in a promise.
+    auto wrapValInPromise = [](jsg::Lock& js, v8::Local<v8::Value> value) {
+      auto resolver = jsg::check(v8::Promise::Resolver::New(js.v8Context()));
+      auto promise = resolver->GetPromise();
+
+      jsg::check(resolver->Resolve(js.v8Context(), value));
+      return js.toPromise(kj::mv(promise));
+    };
+
+    kj::Maybe<jsg::Promise<jsg::Value>> eventualResult;
+    bool caughtException = false;
+    js.tryCatch([&] {
+      eventualResult = wrapValInPromise(js, invokeFn(js, fn, serializedArgs));
+    }, [&](jsg::Value exception) {
+      // We threw while executing JS.
+      caughtException = true;
+      eventualResult = wrapValInPromise(js, exception.getHandle(js));
+    });
+
+    auto& result = KJ_REQUIRE_NONNULL(eventualResult);
+    return processAsyncResult(js, kj::mv(result)).then(js,
+        [caughtException](jsg::Lock& js, SerializedResult serializedResult) -> SerializedResult {
+          if (caughtException) {
+            KJ_IF_SOME(ser, serializedResult.tryGet<Returned>()) {
+              // Since we threw while executing the function, we still want to throw the
+              // serialized result even though serialization was successful.
+              return ThrewException { .serializedException = kj::mv(ser.serializedValue) };
+            }
+          }
+          return kj::mv(serializedResult);
+        });
+  }
+
+  // The JS method returned a promise and we need to wait for the promise to resolve or reject
+  // before serializing the result.
+  jsg::Promise<SerializedResult> processAsyncResult(
```

Comment:
I wonder if it would be clearer at this point to inline this method into its caller?

Reply:
https://github.com/cloudflare/workerd/compare/7562375fb4e0c84af22da84426f7e3097dd4758f..89f20088867ec59912cfe2f8ea7b47e13fae9d7b